prov/gni/src/gnix_mr.c

```diff
@@ -317,43 +397,168 @@ static inline int __mr_cache_entry_get(
  */
 static inline int __mr_cache_entry_put(
 		gnix_mr_cache_t       *cache,
-		gnix_mr_cache_entry_t *entry,
-		RbtIterator           iter)
+		gnix_mr_cache_entry_t *entry)
 {
+	int ret;
 	RbtStatus rc;
+	RbtIterator iter;
+	gnix_mr_cache_key_t *c_key;
 	gni_return_t grc = GNI_RC_SUCCESS;
+	RbtIterator found;
+	gnix_mr_cache_entry_t *c_entry, *parent = NULL;
+	struct dlist_entry *next;
 
 	if (atomic_dec(&entry->ref_cnt) == 0) {
-		rbtErase(cache->inuse, iter);
-		atomic_dec(&cache->inuse_elements);
+		next = entry->siblings.next;
+		dlist_remove(&entry->children);
+		dlist_remove(&entry->siblings);
+
+		/* if this is the last child to deallocate,
+		 * release the reference to the parent
+		 */
+		if (next != &entry->siblings && dlist_empty(next)) {
+			parent = container_of(next, gnix_mr_cache_entry_t,
+					children);
 
-		if (cache->attr.lazy_deregistration) {
-			GNIX_INFO(FI_LOG_MR, "moving key %llu:%llu to stale\n",
+			if (atomic_get(&parent->ref_cnt) == 1) {
+				iter = rbtFind(cache->inuse.rb_tree,
+						(void *) &parent->key);
+				assert(iter);
+
+				rbtErase(cache->inuse.rb_tree, iter);
+			}
+
+			grc = __mr_cache_entry_put(cache, parent);
+			if (unlikely(grc != GNI_RC_SUCCESS)) {
+				GNIX_ERR(FI_LOG_MR,
+						"failed to release reference to parent, "
+						"parent=%p refs=%d\n",
+						parent,
+						atomic_get(&parent->ref_cnt));
+			}
+		}
+
+		atomic_dec(&cache->inuse.elements);
+
+		/* if we are doing lazy dereg and the entry
+		 * isn't retired, put it in the stale cache
+		 */
+		if (cache->attr.lazy_deregistration &&
+				!(entry->flags & GNIX_CE_RETIRED)) {
```

Comment:
I'm a little confused as to when we'd see GNIX_CE_RETIRED here.  What removes it from the cache when it's retired?